pkg/agent/controller/serviceexternalip/controller.go

```diff
@@ -0,0 +1,451 @@
+// workqueue.
+func (c *ServiceExternalIPController) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s", controllerName)
+	defer klog.Infof("Shutting down %s", controllerName)
+
+	go c.localIPDetector.Run(stopCh)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.serviceListerSynced, c.endpointsListerSynced, c.localIPDetector.HasSynced) {
+		return
+	}
+
+	c.removeStaleExternalIPs()
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+// removeStaleExternalIPs unassigns stale external IPs that shouldn't be present on this Node.
+// This function will only delete IPs which caused by Service changes when the agent on this Node was
+// not running. Those IPs should be deleted caused by migration will be deleted by processNextWorkItem.
+func (c *ServiceExternalIPController) removeStaleExternalIPs() {
+	desiredExternalIPs := sets.NewString()
+	services, _ := c.serviceLister.List(labels.Everything())
+	for _, service := range services {
+		if service.Spec.Type == corev1.ServiceTypeLoadBalancer &&
+			service.ObjectMeta.Annotations[types.ExternalIPPoolAnnotationKey] != "" &&
+			len(service.Status.LoadBalancer.Ingress) != 0 {
+			desiredExternalIPs.Insert(service.Status.LoadBalancer.Ingress[0].IP)
+		}
+	}
+	actualExternalIPs := c.ipAssigner.AssignedIPs()
+	for ip := range actualExternalIPs.Difference(desiredExternalIPs) {
+		if err := c.ipAssigner.UnassignIP(ip); err != nil {
+			klog.ErrorS(err, "Failed to clean up stale service external IP", "ip", ip)
+		}
+	}
+}
+
+// worker is a long-running function that will continually call the processNextWorkItem function in
+// order to read and process a message on the workqueue.
+func (c *ServiceExternalIPController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *ServiceExternalIPController) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+	if key, ok := obj.(apimachinerytypes.NamespacedName); !ok {
+		c.queue.Forget(obj)
+		klog.Errorf("Expected NamespacedName in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncService(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.Errorf("Error syncing Service %s, requeuing. Error: %v", key, err)
+	}
+	return true
+}
+
+func (c *ServiceExternalIPController) deleteService(service apimachinerytypes.NamespacedName) error {
+	c.externalIPStatesMutex.Lock()
+	defer c.externalIPStatesMutex.Unlock()
+	if state, exist := c.externalIPStates[service]; !exist {
+		return nil
+	} else {
+		delete(c.externalIPStates, service)
+		return c.ipAssigner.UnassignIP(state.ip)
```

Comment:
No need to indent the block.

Should it unassignIP first and only delete the state on success?